ted/src/main/java/frc/robot/subsystems/CameraSubsystem.java

```diff
@@ -45,18 +56,43 @@ public CameraSubsystem() {
    * @return a vision measurement of the bot pose in field space
    */
   public VisionMeasurement getVisionBotPose(){
-    double tagId = table.getEntry("tid").getDouble(0);
+    double tagId = this.table.getEntry("tid").getDouble(noTagInSightId);
+    NetworkTableEntry botposeEntry = this.table.getEntry(botPoseSource);
+    VisionMeasurement visionMeasurement = new VisionMeasurement(null, 0.0);
 
-    if (tagId == noTagInSightId){
-      return new VisionMeasurement(null, 0.0);
-    }
-    else{
-      double[] botpose = this.table.getEntry("botpose").getDoubleArray(new double[this.defaultDoubleArraySize]);
-      Double timestamp = Timer.getFPGATimestamp() - (botpose[this.TimestampIndex]/this.milisecondsInSeconds);
-      Translation2d botTranslation = new Translation2d(botpose[this.botPositionXIndex], botpose[this.botPositionYIndex]);
+    if (botposeEntry.exists() && tagId != noTagInSightId){
+      double[] botpose = botposeEntry.getDoubleArray(new double[this.BotposeDoubleArraySize]);
+      Double timestamp = (botposeEntry.getLastChange() / this.microsecondsInSeconds) - (botpose[this.latencyIndex]/this.milisecondsInSeconds);
+      Translation2d botTranslation = new Translation2d(botpose[this.fieldSpaceXIndex], botpose[this.fieldSpaceYIndex]);
       Rotation2d botYaw = Rotation2d.fromDegrees(botpose[this.botRotationIndex]);
       Pose2d realRobotPosition = new Pose2d(botTranslation, botYaw);
-      return new VisionMeasurement(realRobotPosition, timestamp);
+      visionMeasurement = new VisionMeasurement(realRobotPosition, timestamp);
+    }
+    return visionMeasurement;
+  }
+
+  /**
+   * a method that gets botPoseSource
+   * @return which limelight datatable we are using
+   */
+  public String getBotPoseSource(){
+    return botPoseSource;
+  }
+
+  /**
+   * a method that sets which limelight data table we should be using
+   * based on the alliance spit out from driver station
+   */
```

Comment:
link to the documentation here about what these different data tables are:
https://docs.limelightvision.io/docs/docs-limelight/apis/complete-networktables-api#apriltag-and-3d-data

Reply:
added